src/overlay005/vs_seeker.c

```diff
@@ -0,0 +1,903 @@
+    case VS_SEEKER_STATE_NO_BATTERY:
+        missingBattery = (VS_SEEKER_MAX_BATTERY - Events_GetVsSeekerBattery(vsSeeker->events));
+
+        if ((missingBattery / 10) == 0) {
+            numDigits = 1;
+        } else if ((missingBattery / 100) == 0) {
+            numDigits = 2;
+        } else {
+            numDigits = 3;
+        }
+
+        StringTemplate_SetNumber(vsSeeker->template, 0, missingBattery, numDigits, 1, 1);
+        VsSeekerSystem_SetState(vsSeeker, 9);
+        break;
+    case VS_SEEKER_STATE_NO_TRAINERS:
+        VsSeekerSystem_SetState(vsSeeker, 9);
+        break;
+    case VS_SEEKER_STATE_DONE:
+        Heap_FreeToHeap(vsSeeker);
+        return TRUE;
+    }
+
+    return FALSE;
+}
+
+static void VsSeekerSystem_SetState(VsSeekerSystem *vsSeeker, enum VsSeekerState state)
+{
+    vsSeeker->state = state;
+}
+
+static enum VsSeekerUsability VsSeekerSystem_CheckUsability(VsSeekerSystem *vsSeeker)
+{
+    if (Events_GetVsSeekerBattery(vsSeeker->events) == VS_SEEKER_MAX_BATTERY) {
+        if (vsSeeker->numVisibleTrainers == 0) {
+            return VS_SEEKER_USABILITY_NO_TRAINERS;
+        }
+
+        return VS_SEEKER_USABILITY_OK;
+    }
+
+    return VS_SEEKER_USABILITY_NO_BATTERY;
+}
+
+static void VsSeekerSystem_CollectViableNpcs(VsSeekerSystem *vsSeeker)
+{
+    // Can't get this to match with C99 style declarations
+    int trainerX, trainerZ;
+    int numVisibleTrainers;
+    int xMin, xMax, zMin, zMax;
+    u32 npcCount = FieldSystem_GetNpcCount(vsSeeker->fieldSystem);
+
+    numVisibleTrainers = 0;
+
+    for (int i = 0; i < npcCount; i++) {
+        vsSeeker->trainers[i] = NULL;
+    }
+
+    int playerX = Player_GetXPos(vsSeeker->fieldSystem->playerAvatar);
+    int playerZ = Player_GetZPos(vsSeeker->fieldSystem->playerAvatar);
+    xMin = playerX - VS_SEEKER_SEARCH_RADIUS_LEFT;
+    xMax = playerX + VS_SEEKER_SEARCH_RADIUS_RIGHT;
+    zMin = playerZ - VS_SEEKER_SEARCH_RADIUS_UP;
+    zMax = playerZ + VS_SEEKER_SEARCH_RADIUS_DOWN;
+
+    if (xMin < 0) {
+        xMin = 0;
+    }
+
+    if (zMin < 0) {
+        zMin = 0;
+    }
+
+    for (int i = 0; i < npcCount; i++) {
+        MapObject *mapObj = MapObjMan_LocalMapObjByIndex(vsSeeker->fieldSystem->mapObjMan, i);
+
+        if (mapObj == NULL) {
+            continue;
+        }
+
+        u32 eventType = MapObject_GetEventType(mapObj);
+
+        switch (eventType) {
+        case 0x1:
+        case 0x2:
+        case 0x4:
+        case 0x5:
+        case 0x6:
+        case 0x7:
+        case 0x8:
+            trainerX = MapObject_GetXPos(mapObj);
+            trainerZ = MapObject_GetZPos(mapObj);
+
+            if ((trainerX >= xMin) && (trainerX <= xMax) && (trainerZ >= zMin) && (trainerZ <= zMax)) {
+                if (VsSeeker_IsMoveCodeHidden(MapObject_GetMoveCode(mapObj)) == FALSE) {
+                    vsSeeker->trainers[numVisibleTrainers] = mapObj;
+                    numVisibleTrainers++;
+                }
+            }
+        }
```

Comment:
A little annoying that the ASM for this looks is a jump-table and that you can't reduce this to an if-block.